### PR TITLE
add linear regression

### PR DESCRIPTION
**Pull request by 123malin**
add linear regression

### CODE REVIEW

File: models/rank/linear_regression/model.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            input=self.predict, label=self.label)
+        avg_cost = fluid.layers.reduce_sum(cost)
+
+        self._cost = avg_cost
+
+        self._metrics["COST"] = self._cost
+        self._metrics["Predict"] = self.predict
+        if is_infer:
+            self._infer_results["Predict"] = self.predict
+            self._infer_results["COST"] = self._cost
```
Comment (seiriosPlus) [inline models/rank/linear_regression/model.py:77]: is `cost` necessary for `inference`?

File: models/rank/linear_regression/data/preprocess.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    for i in range(1, num + 1):
+        with open(os.path.join(output_dir, "part_" + str(i)), 'w') as fout:
+            data = contents[(i - 1) * lines_per_file:min(i * lines_per_file,
+                                                         len(contents))]
+            for line in data:
+                fout.write(line)
+
+
+if __name__ == "__main__":
+    random.seed(1111111)
```
Comment (seiriosPlus) [inline models/rank/linear_regression/data/preprocess.py:140]: is this right?

Comment (seiriosPlus) [inline models/rank/linear_regression/data/preprocess.py:140]: is this right?

File: models/rank/linear_regression/data/preprocess.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            key = "%s:%s" % (arr[0], item)
+            out_str += "%s " % (to_hash(key))
+    return out_str.strip()
+
+
+def get_hash(path):
+    #0-34831 1-time:974673057 2-userid:2021 3-gender:M 4-age:25 5-occupation:0 6-movieid:1345  7-title:Carrie (1976)  8-genres:Horror  9-label:2
+    for line in open(path):
+        arr = line.strip().split("\t")
+        out_str = "logid:%s %s %s %s %s %s %s %s %s %s" % \
```
Comment (seiriosPlus) [inline models/rank/linear_regression/data/preprocess.py:117]: more elegant grammar？

File: models/rank/linear_regression/data/preprocess.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+    res = []
+    for line in open(path):
+        line = line.strip()
+        arr = line.split("::")
+        userid = arr[0]
+        movieid = arr[1]
+        out_str = "time:%s\t%s\t%s\tlabel:%s" % (arr[3], user_dict[userid],
+                                                 movie_dict[movieid], arr[2])
+        log_id = hash(out_str) % 1000000000
```
Comment (seiriosPlus) [inline models/rank/linear_regression/data/preprocess.py:45]: one billion?

File: models/rank/linear_regression/model.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    def net(self, inputs, is_infer=False):
+        init_value_ = 0.1
+        is_distributed = True if envs.get_trainer() == "CtrTrainer" else False
+
+        # ------------------------- network input --------------------------
+
+        sparse_var = self._sparse_data_var
+        self.label = self._dense_data_var[0]
+
+        def embedding_layer(input):
```
Comment (fuyinno4) [inline models/rank/linear_regression/model.py:41]: 函数挪外面，和net（）并列

**Review by MrChengmo — APPROVED** [review]
LGTM

**Pull request closed without merging**